### PR TITLE
Fixed buffer update in BatchNorm when track_running_stats is set to False

### PR DESCRIPTION
**Pull request by frgfm**
This PR aims at tackling #37823 by:
- ensuring that buffers will be used for normalization computation but won't be updated, when buffers are not None, and `track_running_stats=False`
- adding a corresponding unittest to ensure expected behaviour

Any feedback is welcome!

_Note: we might want to update the docstrings of  `BatchNorm*d`, feel free to share any suggestion!_

### CODE REVIEW

File: torch/nn/modules/batchnorm.py
Comment (SsnL) [inline torch/nn/modules/batchnorm.py:106]: maybe you mean "passed *as* None"?

**Review by ssnl — COMMENTED** [review]
Also could you update the docs to reflect this behavior. Currently it says batch stats are always used if track_running_stats is False

File: torch/nn/modules/batchnorm.py
Comment (frgfm) [reply torch/nn/modules/batchnorm.py]: Indeed, fixed the typo, thanks! I don't know why flake fails though.

**Comment by frgfm** [comment]
> Also could you update the docs to reflect this behavior. Currently it says batch stats are always used if track_running_stats is False

Yes, I made a suggestion to change the corresponding docstrings. Is it clear enough in your opinion?

**Comment by SsnL** [comment]
@frgfm Hi, sorry for the late reply. The last weekend has been a bit crazy. 

Upon thinking about this more, I become doubtful whether this fix is the correct behavior. To me, as its name suggests, `track_running_stats` should be controlling whether the running stats are updated or not, but not which stats should be used for normalization. Maybe the expected behavior could be that 

|                                             | train                                      | eval                        |   supported use case |
|---------------------------------------------|--------------------------------------------|-----------------------------|-----------|
| track_running_stats=False && buffer is not None | normalize with batch stats   (no update)              | normalize with buffer stats   (no update)  | N |

**Comment by frgfm** [comment]
Mmmh, so we would end up with


  | train | eval | supported use case
-- | -- | -- | --
track_running_stats=True && buffer is not None | normalize & update buffer with batch stats | normalize with buffer stats | Y
track_running_stats=True && buffer is None | normalize with batch stats | ERROR | N
track_running_stats=False && buffer is not None | normalize with batch stats | normalize with buffer stats | Y
track_running_stats=False && buffer is None | normalize with batch stats | normalize with batch stats | Y

I agree that the expected behaviour would be more logical. To make sure we tackle the initial issue of  freezing running stat updates:

- **Case 1: requires_grad=True**
training=True, track_running_stats=True, buffer is not None. 
normalize & update buffer with batch stats

- **Case 2: requires_grad is switched to False**
training=True, track_running_stats=True, buffer is not None. 

To freeze buffer update in Case 2, I have to switch track_running_stats=False if we move forward with this. Now, my question would be:
is the behaviour difference logical between "track_running_stats=True && buffer is not None && training=False" and "track_running_stats=False && buffer is not None && training=True" ?

I'm asking because currently, the only workaround I found was to switch every BN to eval mode. But I realize this will not have the same effect. Technically, it would be nice not to have to convert each BN to a `FrozenBatchNorm2d` (from `torchvision`)

What do you think @SsnL ?

**Comment by frgfm** [comment]
Hi @SsnL,
Thinking back about this, there are two things I'm not sure how to handle:
- we may be trying to solve two different issues, even though I feel like they are closely connected. Should we split up the PR in two?
- back to the initial problem of #37823, I'm still confused about the expected behaviour. If you were to freeze a group of conv and BN layers, I do feel like not updating the stats should be the default behaviour (perhaps mistakingly). In this specific case, two questions arise in my opinion: should the normalization be done with the buffer or the batch stats? If it's the buffer, how do we switch the layer to that behaviour? (up until now, most of the workarounds I've seen and what I've used is to switch all BN that are frozen to eval mode, meaning normalization with buffer stats).

What do you think we should move forward with?

**Comment by SsnL** [comment]
@frgfm Sorry again for the late reply. Just to make sure that I understand you clearly. Is your original request about freezing and using the running stats (still having trainable affine weights and biases)? I am asking because `FrozenBatchNorm2d` in `torchvision` also freeze those as well https://github.com/pytorch/vision/blob/39021408587eb252ebb842a54195d840d6d76095/torchvision/ops/misc.py#L43. 

I agree that there are two separate issues here, i.e.,
1. `forward` should not check for nullity.
2. (if my understanding above is correct) how to ``freeze'' a BN layer's batch stats without messing with the `eval` and `train` switches. 

As you said, 1 is straightforward, and 2, not so much. I went to check what tf does. In their case, for `v1`, they have the same behavior as in PyTorch, where one would have to switch between training and inference manually. In `v2`, on setting `layer.trainable = False` (which I believe is the same as `for p in model.parameters(): p.requires_grad_(False)`, they also change the layer to inference mode, for the exact case of finetuning / transferring.

Personally I don't think their `v2` way, which special-cases a single layer on a common API, is a good idea. Yet, without adding another API / flag, the only way seems to write a helper function that flips between train/eval while keeping BN layers in eval.

**Comment by frgfm** [comment]
No worries @SsnL, I understand you had a lot on your plate with your recent work on Contrastive Representation Learning haha (congrats btw)

Bach to your question:
- initially, I was only looking for the best way to use `BatchNorm2d` in frozen layers (in the case of transfer learning). Obviously, `FrozenBatchNorm2d` was a good option but required replacing all BNs in a model (and there was no `eps` in frozenBN until very recently). I went looking for a simple flag that would work, and naively in the first place, I experimented with `track_running_stats`, which yielded the strange behaviour I pointed out in #37823.
- the point of this PR is indeed to fix the behaviour as shown in our previous comments here.
- yet, if we can also find a way to also easy change in BN behaviour when they need to apply for both train and eval modes, that would be great!

Anyway, I agree that specializing the API for one layer is a bad idea. However, to the best of my understanding, the train/eval switch is only here to allow a conditional behaviour that only depends on whether this layer is to be trained or not. If this is correct, then for BN, intuitively, I would say:
- we need to dynamically be able to tell the layer whether it should update its population stats' estimators (buffer) or not
- we also need to dynamically be able to tell the layer whether the stats used for normalization should be the actual current mini-batch stats or the already seen population's estimators (buffer).
- the train --> eval switch should use the computed population mean & variance for normalization, prevent population estimators from updating, and avoid updating the affine parameters.

Thus 2 boolean attributes, whose value cannot be changed by the train/eval switch ( `update_running_stats` & `minibatch_normalization` attributes for instance). Since, in the current API, switching to train/eval mode does not change the parameters' `requires_grad` then:
- updating population stats (buffer) is only done if `self.update_running_stats and self.training`
- normalization is done with mini-batch stats rather than buffer's if `self.minibatch_normalization or self.training`

I'd suggest default values being `update_running_stats=True` and `minibatch_normalization=False`! This might be a big change for another PR, but what are your thoughts on this?

And should we move forward with the current PR after I made the suggested modification you made earlier? (cf. below)

train | eval | supported use case
-- | -- | --
track_running_stats=False && buffer is not None | normalize with batch stats (no update) | normalize with buffer stats (no update) | N

**Comment by SsnL** [comment]
@frgfm Thanks :)

I see. The usecase makes a lot of sense. IIUC, the `minibatch_normalization` really means something like `minibatch_normalization_in_eval`, right? If it is False, the layer would use buffer stats, if available (i.e., not None). Do you want to open an issue to discuss this and see if other maintainers have comments? :)

Yes, as for this PR, I think we should go for the behavior as listed in your table (assuming that the header and the row is misaligned for 1 column).

Thanks for giving this much thoughts!

**Comment by frgfm** [comment]
Correct @SsnL :ok_hand: 
I fixed the table for further references!
And I just made the modifications for this PR and opened the issue #39674

Cheers!

**Comment by SsnL** [comment]
Hmm pytorchbot rebase doesn't seem to work anymore. Could you rebase to master please? :) thanks!

**Comment by frgfm** [comment]
@SsnL sure! Just to make sure, if I rebase master into this branch, I'll have to force push the update. Is that OK for the bot / merge afterwards?

**Comment by SsnL** [comment]
Yes force push is fine!

**Comment by frgfm** [comment]
@SsnL here you go :ok_hand:

**Review by ssnl — APPROVED** [review]
LGTM, thank you! The CI failures aren't your fault.

**Comment by seyeeet** [comment]
I wonder if this issue got solved and we can finally expect to have the following?
<html>
<body>
<p style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; color: rgb(36, 41, 46); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"></p>

  | train | eval | supported use case
-- | -- | -- | --
track_running_stats=False && buffer is not None | normalize with batch stats (no update) | normalize with buffer stats (no update) | N --> Y

<p style="box-sizing: border-box; margin-top: 0px; margin-bottom: 0px !important; color: rgb(36, 41, 46); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"></p>
</body>
</html>

**Comment by frgfm** [comment]
Hi @seyeeet :wave: 

As of PyTorch 1.8, with the following snippet:
```python
import torch
import torch.nn as nn

bn = nn.BatchNorm1d(4)
bn.track_running_stats = False

# Store initial values
num_batches = bn.num_batches_tracked.clone() 
running_mean = bn.running_mean.clone() 
running_var = bn.running_var.clone()

# Forward random tensor in training mode
t = torch.rand(32, 4)
v1 = bn(t)

# Check which stats were updated
print("Training mode")
print(torch.equal(num_batches, bn.num_batches_tracked))
print(torch.equal(running_mean, bn.running_mean))
print(torch.equal(running_var, bn.running_var))

# Forward random tensor in eval mode
bn = bn.eval()
v2 = bn(t)

# Check which stats were updated
print("Eval mode")
print(torch.equal(num_batches, bn.num_batches_tracked))
print(torch.equal(running_mean, bn.running_mean))
print(torch.equal(running_var, bn.running_var))

# Check whether normalization was performed with the same stats
print("Normalization with the same stats in both mode?")
print(torch.equal(v1, v2))
```

you'll get:
```
Training mode
True
True
True
Eval mode
True
True
True
Normalization with the same stats in both mode?
False
```

So yes, it was solved :+1:

**Comment by seyeeet** [comment]
@frgfm thanks for the information and example. it is very helpful.
But I am a little confused and appreciate if you can clear things for me please.
Im using pytorch 1.9

Lets say I have a pretrained model and I want to freeze the BNs, one way is by using `requires_grad=False` and then apply `eval()` to the BNs.
When I was reading this issue I thought I can freeze the BNs by a hacky way as follows (can I?):
I replace all the BNs in my model with new BNs while making the `afffine=False` and then replace the parameters of the new BNs based on their corresponding BNs from the pretrained model. In this way I dont have to send the BN of the model to `eval()` every time, but of course I have to do the `requires_grad=False`, but I am not sure if I should have the `track_running_stats ` True or False.

e.g.

lets consider `bn_0` as the bn of a pretrained model and `bn_2` as the replaced one, so I do:

```
import torch
import torch.nn as nn

# get the params and buffers of the pretrained bn (bn_1)
bn_0 = nn.BatchNorm1d(4,affine=True)
params = list(bn_0.named_parameters()).copy()
buffers = list(bn_0.named_buffers()).copy()
    
    
bn_2 = nn.BatchNorm1d(4,affine=False)
bn_2.track_running_stats = True # not sure if I should set this True or False

bn_2.weight = params[0][1]
bn_2.bias = params[1][1]
bn_2.running_mean = buffers[0][1]
bn_2.running_var = buffers[1][1]
bn_2.num_batches_tracked = buffers[2][1]
bn_2.weight.requires_grad = False
bn_2.bias.requires_grad = False

# Store initial values
num_batches = bn_2.num_batches_tracked.clone() 
running_mean = bn_2.running_mean.clone() 
running_var = bn_2.running_var.clone()



# Forward random tensor in training mode
t = torch.rand(32, 4)
v1 = bn_2(t)

# Check which stats were updated
print("Training mode")
print(torch.equal(num_batches, bn_2.num_batches_tracked))
print(torch.equal(running_mean, bn_2.running_mean))
print(torch.equal(running_var, bn_2.running_var))

# Forward random tensor in eval mode
bn_2 = bn_2.eval()
v2 = bn_2(t)

# Check which stats were updated
print("Eval mode")
print(torch.equal(num_batches, bn_2.num_batches_tracked))
print(torch.equal(running_mean, bn_2.running_mean))
print(torch.equal(running_var, bn_2.running_var))

# Check whether normalization was performed with the same stats
print("Normalization with the same stats in both mode?")
print(torch.equal(v1, v2))
```

another confusing point is that, based on the table

<body>


track_running_stats=True && buffer is not None | normalize & update buffer with batch stats | normalize with buffer stats | Y
-- | -- | -- | --



</body>
</html>

shoudnot 

```
import torch
import torch.nn as nn

bn = nn.BatchNorm1d(4)
bn.track_running_stats = True


# Forward random tensor in training mode
t = torch.rand(32, 4)
v1 = bn(t)


# Forward random tensor in eval mode
bn = bn.eval()
v2 = bn(t)



# Check whether normalization was performed with the same stats
print("Normalization with the same stats in both mode?")
print(torch.equal(v1, v2))
```
give us `True` for `print(torch.equal(v1, v2))`?

**Comment by frgfm** [comment]
Freezing a BN would conventionally mean that it doesn't learn its affine parameters or update its distribution stats. However, the way the incoming data will be normalized during a forward is up to you. By default, the behaviour is just like before my PR.

My PR ensures that modifying the BN attributes will dynamically change both the stat updating & data normalizing behaviours.
Straight to your question: I would suggest keeping `affine=True`, but making you change `track_running_stats=False` after instantiating it.

Regarding your last question, it only involves which data is used for normalization (& whether affine parameters are present, but let's assume you kept the default). The whole point of my PR is that switching between training & evaluation mode changes the normalization behaviour (as per the table in this very PR).

Let me know if some things aren't very clear!

**Comment by alexander-soare** [comment]
@frgfm thanks for all the detailing of the progress on this one. I noticed @SsnL had some feedback on how to make the functionality true to the semantics of the variable names (https://github.com/pytorch/pytorch/pull/38084#issuecomment-627479384)

I think the idea is correct, but then I don't see how it could be useful. What's a scenario in which you would want to train with batch stats and eval with buffer stats?

I noticed this after trying to set `track_running_stats` to False in order to freeze batch norm for fine-tuning + detection and seeing an explosion in my val loss while my train loss stayed low (which makes sense actually - we are using different norm stats for train/eval). Prior to digging into this PR, I was under the impression that I was doing the equivalent of [FrozenBatchNorm2d](https://github.com/pytorch/vision/blob/7947fc8fb38b1d3a2aca03f22a2e6a3caa63f2a0/torchvision/ops/misc.py#L45). But really, that's doing another alternative that's not in @frgfm 's table https://github.com/pytorch/pytorch/pull/38084#issuecomment-627556108. So to clarify, FrozenBatchNorm2d is doing "train with buffer, val with buffer".

So firstly, am I confused about anything I've raised above?

Secondly, if I'm not confused/wrong, I would say that although the behaviour of `BatchNorm` is now semantically correct, it's still not useful. I'd then want a whole new option called `use_buffer_stats`. And to "freeze" batch norm as is done in `FrozenBatchNorm2d` I would set `use_buffer_stats = True` and `track_running_stats = False`. Then finally if you are still with me and agree by now, might as well add a special `.freeze` method to BatchNorm to avoid confusion among anyone who knows via common wisdom that "freezing" batch norm is useful, but hasn't dug deep into exactly what's happening in PyTorch's BatchNorm.

**Comment by frgfm** [comment]
Hello there :wave: 

> I think the idea is correct, but then I don't see how it could be useful. What's a scenario in which you would want to train with batch stats and eval with buffer stats?

There are two parts of the answer:
- some cases are just defined because one class attribute has enough values to create a new combination. Leaving this combination as undefined is a bit dangerous
- the starting point of this topic was about freezing parts of the models that include batch norm. We are just providing some options here, but the "best" choice would probably require a whole research paper :sweat_smile: 


> I noticed this after trying to set `track_running_stats` to False in order to freeze batch norm for fine-tuning + detection and seeing an explosion in my val loss while my train loss stayed low (which makes sense actually - we are using different norm stats for train/eval). Prior to digging into this PR, I was under the impression that I was doing the equivalent of [FrozenBatchNorm2d](https://github.com/pytorch/vision/blob/7947fc8fb38b1d3a2aca03f22a2e6a3caa63f2a0/torchvision/ops/misc.py#L45). But really, that's doing another alternative that's not in @frgfm 's table [#38084 (comment)](https://github.com/pytorch/pytorch/pull/38084#issuecomment-627556108). So to clarify, FrozenBatchNorm2d is doing "train with buffer, val with buffer".
> 
> So firstly, am I confused about anything I've raised above?

You are correct, but this can only be similar to FrozenBN when set in the specific mode (combination of class attributes' values). By explosion, do you mean compared to previous trainings? or compared to previous epochs in the same training?



> **Secondly, if I'm not confused/wrong, I would say that although the behaviour of `BatchNorm` is now semantically correct, it's still not as useful as it could be**. I'd then want a whole new option called `use_buffer_stats`. And to "freeze" batch norm as is done in `FrozenBatchNorm2d` I would set `use_buffer_stats = True` and `track_running_stats = False`. Then finally if you are still with me and agree by now, might as well add a special `.freeze` method to BatchNorm to avoid confusion among anyone who knows via common wisdom that "freezing" batch norm is useful, but hasn't dug deep into exactly what's happening in PyTorch's BatchNorm.

So I agree that this should be simpler, but we cannot ignore nor break the behaviour of `torch.nn.Module` with the `training` attribute and the train/eval mode. That would most likely require many discussions with the PyTorch team to implement this but I agree that this would be useful!

**Comment by alexander-soare** [comment]
@frgfm thanks for picking this up. Some responses and clarifications for everyone's benefit

> the "best" choice would probably require a whole research paper

Just want to reiterate that I'm aiming for the choice that PyTorch, Detectron2, and other libraries seem to be using to "freeze" batch norm for the purpose of fine-tuning on downstream tasks. If there are many other common ways that one would "freeze" a batch norm, then I suppose my ask is not so straight-forward.

> By explosion, do you mean...

"Explosion" was probably the wrong term as I'm not talking about the loss having started off in a nice place then exploding. I get good train loss but the val loss is out-of-the-ballpark huge (from the very first validation). But I'm not surprised. This is to be expected given the current (post-PR) behaviour of BatchNorm. Training with `track_running_stats = True` doesn't have this issue. And converting BatchNorm2d to [FrozenBatchNorm2d](https://github.com/pytorch/vision/blob/7947fc8fb38b1d3a2aca03f22a2e6a3caa63f2a0/torchvision/ops/misc.py#L45) also doesn't have this issue.

> but this can only be similar to FrozenBN when set in the specific mode

Hmm, well I'm saying that going by the table [here](https://github.com/pytorch/pytorch/pull/38084#issuecomment-627556108), it seems there's no way to do it. If I'm mistaken though, then my qualm is solved :)

> but we cannot ignore nor break the behaviour of torch.nn.Module

Yeah you're right. I don't want my `.freeze` suggestion, or even my particular suggestions for parameter names to detract from my **main point which is**: _**BatchNorm should have straight-forward way of doing what TorchVision's FrozenBatchNorm2d does**_. 

> I agree that this would be useful!

Cool :) Waiting on feedback from someone else then might open a new issue.

**Pull request closed without merging**
